File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Sprite.Node/LNSprite.cs
using Loon.Core.Graphics.Opengl;
using Loon.Core.Geom;
using System;
using Loon.Core;
using Loon.Utils;
namespace Loon.Action.Sprite.Node {
	
	public class LNSprite : LNNode {
	
		private BlendState blendState = BlendState.NonPremultiplied;
	
		private float rotation;
	
		protected internal LTexture _texture;
	
		private float[] pos, scale;
	
		protected internal LNAnimation _ans;
	
		protected internal bool _flipX = false, _flipY = false;
	
		protected internal System.Collections.Generic.Dictionary<string, LNAnimation> _anims;
	
		public LNSprite(RectBox rect):base(rect) {
			this._ans = null;
			base._anchor = new Vector2f(0f, 0f);
		}
	
		public LNSprite(int x, int y, int w, int h):base(x,y,w,h) {
			this._ans = null;
			base._anchor = new Vector2f(0f, 0f);
		}

        public LNSprite(string fsName)
            : base()
        {
            this._ans = null;
            LNFrameStruct struc0 = LNDataCache.GetFrameStruct(fsName);
            if (struc0 == null)
            {
                throw new Exception("");
            }
            this._texture = struc0._texture;
            base._left = struc0._textCoords.X();
            base._top = struc0._textCoords.Y();
            base._orig_width = struc0._orig_width;
            base._orig_height = struc0._orig_height;
            base.SetNodeSize(struc0._size_width, struc0._size_height);
            base._anchor.Set(struc0._anchor);
            blendState = struc0._state;
            if (!struc0._place.Equals(0, 0))
            {
                SetPosition(struc0._place);
            }
        }
	
		public LNSprite():this(LSystem.screenRect) {
			
		}
	
		public static LNSprite GInitWithTexture(LTexture tex2d) {
			LNSprite sprite = new LNSprite();
			sprite.InitWithTexture(tex2d);
			return sprite;
		}
	
		public static LNSprite GInitWithFilename(string file) {
			LNSprite sprite = new LNSprite();
			sprite.InitWithFilename(file);
			return sprite;
		}
	
		public static LNSprite GInitWithFrameStruct(LNFrameStruct fs) {
			LNSprite sprite = new LNSprite();
			sprite.InitWithFrameStruct(fs);
			return sprite;
		}
	
		public static LNSprite GInitWithFrameStruct(string fsName) {
			return GInitWithFrameStruct(LNDataCache.GetFrameStruct(fsName));
		}
	
		public static LNSprite GInitWithAnimation(LNAnimation ans) {
			LNSprite sprite = new LNSprite();
			sprite.InitWithAnimation(ans, 0);
			return sprite;
		}
	
		public override void Draw(SpriteBatch batch) {
			if (base._visible && (this._texture != null)) {
				pos = base.ConvertToWorldPos();
				if (_screenRect.Intersects(pos[0], pos[1], GetWidth(), GetHeight())
						|| _screenRect.Contains(pos[0], pos[1])) {
					if (_parent != null) {
						rotation = ConvertToWorldRot();
						scale = ConvertToWorldScale();
					} else {
						rotation = _rotation;
						scale[0] = _scale.x;
						scale[1] = _scale.y;
					}
					batch.SetColor(base._color.r, base._color.g, base._color.b,
							base._alpha);
					if (rotation == 0) {
						batch.Draw(_texture, pos[0], pos[1], base._size_width
								* scale[0], base._size_height * scale[1],
								base._left, base._top, base._orig_width,
								base._orig_height, _flipX, _flipY);
					} else {
						batch.Draw(_texture, pos[0], pos[1], _anchor.x, _anchor.y,
								base._size_width, base._size_height, scale[0],
								scale[1], MathUtils.ToDegrees(rotation),
								base._left, base._top, base._orig_width,
								base._orig_height, _flipX, _flipY);
					}
					batch.ResetColor();
					BlendState oldState = batch.GetBlendState();
					if (blendState != oldState) {
						batch.Flush(blendState);
						batch.SetBlendState(oldState);
					}
				}
			}
		}
	
		public void InitWithTexture(LTexture tex2d) {
			this._texture = tex2d;
			base._left = 0;
			base._top = 0;
			base.SetNodeSize(_texture.GetWidth(),_texture.GetHeight());
			base._anchor.Set(base.GetWidth() / 2f, base.GetHeight() / 2f);
		}
	
		public void InitWithFilename(string filename) {
			this._texture = LTextures.LoadTexture(filename, Loon.Core.Graphics.Opengl.LTexture.Format.LINEAR);
			base._left = 0;
			base._top = 0;
			base.SetNodeSize(_texture.GetWidth(),_texture.GetHeight());
			base._anchor.Set(base.GetWidth() / 2f, base.GetHeight() / 2f);
		}
	
		public void InitWithFrameStruct(LNFrameStruct fs) {
			this._texture = fs._texture;
			blendState = fs._state;
			base._left = fs._textCoords.X();
			base._top = fs._textCoords.Y();
			base._orig_width = fs._orig_width;
			base._orig_height = fs._orig_height;
			base.SetNodeSize(fs._size_width,fs._size_height);
			base._anchor.Set(fs._anchor);
		}
	
		public void InitWithAnimation(LNAnimation ans, int idx) {
			if (ans != null) {
				if (this._anims == null) {
					this._anims = new System.Collections.Generic.Dictionary<string, LNAnimation>();
				}
				this._ans = ans;
				InitWithFrameStruct(this._ans.GetFrame(idx));
				CollectionUtils.Put(_anims,ans.GetName(),ans);
			}
		}
	
		public void AddAnimation(LNAnimation anim) {
			InitWithAnimation(anim, 0);
		}
	
		public void SetFrame(string animName, int index) {
			if (this._anims == null) {
                this._anims = new System.Collections.Generic.Dictionary<string, LNAnimation>();
			}
			if (this._anims.ContainsKey(animName)) {
				this._ans = (LNAnimation)CollectionUtils.Get(this._anims,animName);
				InitWithAnimation(this._ans, index);
			}
		}
	
		public void SetFrame(int idx) {
			if (_ans != null) {
				InitWithFrameStruct(_ans.GetFrame(idx));
			}
		}
	
		public void SetFrameTime(float time) {
			if (_ans != null) {
				InitWithFrameStruct(_ans.GetFrameByTime(time));
			}
		}
	
		public LTexture GetFrameTexture() {
			return this._texture;
		}
	
		public LNAnimation GetAnimation() {
			return this._ans;
		}
	
		public void SetAnimation(LNAnimation ans) {
			this._ans = ans;
			this.InitWithTexture(this._ans.GetFrame(0)._texture);
		}
	
		public bool IsFlipX() {
			return _flipX;
		}
	
		public void SetFlipX(bool flipX) {
			this._flipX = flipX;
		}
	
		public bool IsFlipY() {
			return _flipY;
		}
	
		public void SetFlipY(bool flipY) {
			this._flipY = flipY;
		}
	
		public BlendState GetBlendState() {
			return blendState;
		}
	
		public void SetBlendState(BlendState blendState) {
			this.blendState = blendState;
		}
	}
}

File: Java/Android/OpenGLES/source/core/src/loon/action/sprite/node/LNSprite.java
/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
package loon.action.sprite.node;

import java.util.HashMap;

import loon.action.sprite.SpriteBatch;
import loon.action.sprite.SpriteBatch.BlendState;
import loon.core.LSystem;
import loon.core.geom.RectBox;
import loon.core.geom.Vector2f;
import loon.core.graphics.opengl.LTexture;
import loon.core.graphics.opengl.LTexture.Format;
import loon.core.graphics.opengl.LTextures;
import loon.utils.MathUtils;

public class LNSprite extends LNNode {

	private BlendState blendState = BlendState.NonPremultiplied;

	private float rotation;

	protected LTexture _texture;

	private float[] pos, scale;

	protected LNAnimation _ans;

	protected boolean _flipX = false, _flipY = false;

	protected HashMap<String, LNAnimation> _anims;

	public LNSprite(RectBox rect) {
		super(rect);
		this._ans = null;
		super._anchor = new Vector2f(0f, 0f);
	}

	public LNSprite(int x, int y, int w, int h) {
		super(x, y, w, h);
		this._ans = null;
		super._anchor = new Vector2f(0f, 0f);
	}

	public LNSprite(String fsName) {
		super();
		this._ans = null;
		LNFrameStruct struct = LNDataCache.getFrameStruct(fsName);
		if (struct == null) {
			throw new RuntimeException("");
		}
		this._texture = struct._texture;
		super._left = struct._textCoords.x();
		super._top = struct._textCoords.y();
		super._orig_width = struct._orig_width;
		super._orig_height = struct._orig_height;
		super.setNodeSize(struct._size_width, struct._size_height);
		super._anchor.set(struct._anchor);
		blendState = struct._state;
		if (!struct._place.equals(0, 0)) {
			setPosition(struct._place);
		}
	}

	public LNSprite() {
		this(LSystem.screenRect);
	}

	public static LNSprite GInitWithTexture(LTexture tex2d) {
		LNSprite sprite = new LNSprite();
		sprite.initWithTexture(tex2d);
		return sprite;
	}

	public static LNSprite GInitWithFilename(String file) {
		LNSprite sprite = new LNSprite();
		sprite.initWithFilename(file);
		return sprite;
	}

	public static LNSprite GInitWithFrameStruct(LNFrameStruct fs) {
		LNSprite sprite = new LNSprite();
		sprite.initWithFrameStruct(fs);
		return sprite;
	}

	public static LNSprite GInitWithFrameStruct(String fsName) {
		return GInitWithFrameStruct(LNDataCache.getFrameStruct(fsName));
	}

	public static LNSprite GInitWithAnimation(LNAnimation ans) {
		LNSprite sprite = new LNSprite();
		sprite.initWithAnimation(ans, 0);
		return sprite;
	}

	@Override
	public void draw(SpriteBatch batch) {
		if (super._visible && (this._texture != null)) {
			pos = super.convertToWorldPos();
			if (_screenRect.intersects(pos[0], pos[1], getWidth(), getHeight())
					|| _screenRect.contains(pos[0], pos[1])) {
				if (_parent != null) {
					rotation = convertToWorldRot();
					scale = convertToWorldScale();
				} else {
					rotation = _rotation;
					scale[0] = _scale.x;
					scale[1] = _scale.y;
				}
				batch.setColor(super._color.r, super._color.g, super._color.b,
						super._alpha);
				if (rotation == 0) {
					batch.draw(_texture, pos[0], pos[1], super._size_width
							* scale[0], super._size_height * scale[1],
							super._left, super._top, super._orig_width,
							super._orig_height, _flipX, _flipY);
				} else {
					batch.draw(_texture, pos[0], pos[1], _anchor.x, _anchor.y,
							super._size_width, super._size_height, scale[0],
							scale[1], MathUtils.toDegrees(rotation),
							super._left, super._top, super._orig_width,
							super._orig_height, _flipX, _flipY);
				}
				batch.resetColor();
				BlendState oldState = batch.getBlendState();
				if (blendState != oldState) {
					batch.flush(blendState);
					batch.setBlendState(oldState);
				}
			}
		}
	}

	public final void initWithTexture(LTexture tex2d) {
		this._texture = tex2d;
		super._left = 0;
		super._top = 0;
		super.setNodeSize(_texture.getWidth(), _texture.getHeight());
		super._anchor.set(super.getWidth() / 2f, super.getHeight() / 2f);
	}

	public final void initWithFilename(String filename) {
		this._texture = LTextures.loadTexture(filename, Format.LINEAR);
		super._left = 0;
		super._top = 0;
		super.setNodeSize(_texture.getWidth(), _texture.getHeight());
		super._anchor.set(super.getWidth() / 2f, super.getHeight() / 2f);
	}

	public final void initWithFrameStruct(LNFrameStruct fs) {
		this._texture = fs._texture;
		blendState = fs._state;
		super._left = fs._textCoords.x();
		super._top = fs._textCoords.y();
		super._orig_width = fs._orig_width;
		super._orig_height = fs._orig_height;
		super.setNodeSize(fs._size_width, fs._size_height);
		super._anchor.set(fs._anchor);
	}

	public final void initWithAnimation(LNAnimation ans, int idx) {
		if (ans != null) {
			if (this._anims == null) {
				this._anims = new HashMap<String, LNAnimation>();
			}
			this._ans = ans;
			initWithFrameStruct(this._ans.getFrame(idx));
			_anims.put(ans.getName(), ans);
		}
	}

	public void addAnimation(LNAnimation anim) {
		initWithAnimation(anim, 0);
	}

	public void setFrame(String animName, int index) {
		if (this._anims == null) {
			this._anims = new HashMap<String, LNAnimation>();
		}
		if (this._anims.containsKey(animName)) {
			this._ans = this._anims.get(animName);
			initWithAnimation(this._ans, index);
		}
	}

	public void setFrame(int idx) {
		if (_ans != null) {
			initWithFrameStruct(_ans.getFrame(idx));
		}
	}

	public void setFrameTime(float time) {
		if (_ans != null) {
			initWithFrameStruct(_ans.getFrameByTime(time));
		}
	}

	public final LTexture getFrameTexture() {
		return this._texture;
	}

	public final LNAnimation getAnimation() {
		return this._ans;
	}

	public final void setAnimation(LNAnimation ans) {
		this._ans = ans;
		this.initWithTexture(this._ans.getFrame(0)._texture);
	}

	public boolean isFlipX() {
		return _flipX;
	}

	public void setFlipX(boolean flipX) {
		this._flipX = flipX;
	}

	public boolean isFlipY() {
		return _flipY;
	}

	public void setFlipY(boolean flipY) {
		this._flipY = flipY;
	}

	public BlendState getBlendState() {
		return blendState;
	}

	public void setBlendState(BlendState blendState) {
		this.blendState = blendState;
	}
}
